### PR TITLE
add support for mssql, fix problem with wrong .net framework

### PR DESCRIPTION
**Pull request by kosnkow**

### CODE REVIEW

**Comment by ItsInsajd** [comment]
One more thing, we noticed that tests fail after project is generated, this due to db context change. There are several options to fix this:
- set up tests so that they use sqlite no matter which database has been selected
- mock dbcontext so that no real database is used at all
- set up test connection string for separate sql server db used only for tests 
I think that the third option is obviously bad, because we don't want to set up different database just to run tests. First option is also flawed, because using different databases for test and production may lead to tests not being reliable (something may work on sqlite and fail for sql server). So second option is probably the best. 

Could somebody from the team elaborate more on the topic? I'm trying with @kosnkow to find the best solution for this case, but we don't want to differ too much from existing approach

**Comment by maznag** [comment]
@kosnkow can you add the other databases: Oracle, Postgresql and MySQL or can   we do it in a separate PR ?

**Comment by kosnkow** [comment]
> @kosnkow can you add the other databases: Oracle, Postgresql and MySQL or can we do it in a separate PR ?

separete, now we try to finish mssql but there is problem with tests, they should mock the context

**Comment by kosnkow** [comment]
> Is not normal. Can you give me the error ?

errors
you mean its normal I can't compile solution ?

**Comment by maznag** [comment]
@kosnkow why you have change netstandard2.1 to netcoreapp3.1 in JHipsterNet.csproj.ejs.

The latest version of netstandard is 2.1. See https://docs.microsoft.com/en-us/dotnet/standard/frameworks

Yo should have this in JHipsterNet.csproj.ejs. netstandard2.1 don't netcoreapp:

<PropertyGroup>
        <TargetFramework>netstandard2.1</TargetFramework>
        <RootNamespace>JHipsterNet</RootNamespace>
    </PropertyGroup>

**Comment by kosnkow** [comment]
error

**Comment by kosnkow** [comment]
errors

**Comment by kosnkow** [comment]
Ok target framework reverted, seems this happen only on my local machine

**Comment by danielpetisme** [comment]
This is awesome, we were definitively lacking "true" DB support.

Thank you!!

Let me know if you need help

**Comment by ItsInsajd** [comment]
After some debate we decided to keep tests running on sqlite, so I created separate startup for tests. We ran into several problem with some tests failing on sql server. This is probably due to tests written with sqlite in mind, so not everything works correctly with diffrent DB. 

So right now, no matter which db user chooses, test setup makes them run on in-memory sqlite.

**Comment by kosnkow** [comment]
@maznag  @danielpetisme hi guys we just added oracle support, and that would be all for relational databases, we also plan to add support for mongo , but this is different story since we need to change a lot the template.  Could you merge current changes  ?

File: generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs [hunk 1]
```diff
@@ -16,7 +16,18 @@ using System;
 using <%= namespace %>.Data;
 using Microsoft.AspNetCore.Builder;
 using Microsoft.Extensions.Hosting;
+<%_ switch(databaseType) {
```
Comment (danielpetisme) [inline generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs]: What if the user select Oracle? I don't see the according imports

File: generators/server/prompts.js [hunk 2]
```diff
@@ -31,14 +31,42 @@ function askForServerSideOpts() {
...
             default: 'MyCompany'
+        },
+        {
+            type: 'list',
+            name: 'database',
+            message: 'Wchich database do you want to use',
+            choices: [
+                {
+                    value: 'sqllite',
+                    name: 'sql lite in memory'
```
Comment (danielpetisme) [inline generators/server/prompts.js]: We should respect the product namin/casing
sql lite => SQLite (cf. https://www.sqlite.org/index.html)

File: generators/server/prompts.js [hunk 3]
```diff
@@ -31,14 +31,42 @@ function askForServerSideOpts() {
...
+                },
+                {
+                    value: 'mssql',
+                    name: 'ms sql server'
```
Comment (danielpetisme) [inline generators/server/prompts.js]: Same here ms sql server => Microsoft SQL Server (https://en.wikipedia.org/wiki/Microsoft_SQL_Server)

File: generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs [hunk 4]
```diff
@@ -26,22 +37,55 @@ namespace <%= namespace %>.Infrastructure {
...
         }
 
         public static IApplicationBuilder UseApplicationDatabase(this IApplicationBuilder @this,
             IServiceProvider serviceProvider, IHostEnvironment environment)
         {
-            if (environment.IsDevelopment() || environment.IsProduction()) {
-                var context = serviceProvider.GetRequiredService<ApplicationDatabaseContext>();
-                context.Database.OpenConnection();
-                context.Database.EnsureCreated();
+            if (environment.IsDevelopment() || environment.IsProduction())
```
Comment (danielpetisme) [inline generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs]: How many profiles can we have ?
dev/prod ? what else? the conditions looks weird.

File: generators/server/templates/dotnetcore/src/Project/Infrastructure/IdentityStartup.cs.ejs [hunk 5]
```diff
@@ -25,17 +25,19 @@ namespace <%= namespace %>.Infrastructure {
...
-            var roleManager = serviceProvider.GetRequiredService<RoleManager<Role>>();
+            using (var scope = serviceProvider.CreateScope())
+            {
+                var userManager = scope.ServiceProvider.GetRequiredService<UserManager<User>>();
+                var roleManager = scope.ServiceProvider.GetRequiredService<RoleManager<Role>>();
 
-            //            var op1 = new Operation {Date = DateTime.Now, Description = "An operation", Amount = (decimal) 10.0};
-//            var op2 = new Operation {Date = DateTime.Now, Description = "Another operation", Amount = (decimal) 20.0};
-//            var op3 = new Operation {Date = DateTime.Now, Description = "Last one operation", Amount = (decimal) 30.0};
+                //            var op1 = new Operation {Date = DateTime.Now, Description = "An operation", Amount = (decimal) 10.0};
```
Comment (danielpetisme) [inline generators/server/templates/dotnetcore/src/Project/Infrastructure/IdentityStartup.cs.ejs]: Can be removed, it certainly represents early tests I did with  BankAccount/ Operation domain classes

File: generators/server/templates/dotnetcore/src/Project/Project.csproj.ejs [hunk 6]
```diff
@@ -37,13 +37,38 @@
...
         <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="3.1.1" />
-        <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="3.1.1" />
         <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.2.0" />
         <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.1" />
         <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="3.1.0" Condition="'$(Configuration)' == 'Debug'" />
         <PackageReference Include="Microsoft.AspNetCore.Mvc.Analyzers" Version="2.2.0" />
-        <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.1" />
+        <%_ switch(databaseType) {
+            case 'sqllite': _%>
+        <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="3.1.1" />
```
Comment (danielpetisme) [inline generators/server/templates/dotnetcore/src/Project/Project.csproj.ejs]: We may be able to factorize this dependency (every DB use it), but don't feel obligated. How it behaves with Mongo `Microsoft.AspNetCore.Identity.EntityFrameworkCore` support it too?

Comment (ItsInsajd) [reply generators/server/templates/dotnetcore/src/Project/Project.csproj.ejs]: Mongo is big work in progress for now, I'm trying to use https://github.com/tugberkugurlu/AspNetCore.Identity.MongoDB for identity, we will create another PR if it works out

File: generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs (hunk 4 above)
Comment (ItsInsajd) [reply generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs]: This is not our change, I can't tell you much about this piece of code. I just added scoping for this part, nothing more

File: generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs (hunk 1 above)
Comment (kosnkow) [reply generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs]: Is not needed, Oracle.EntityFrameworkCore adds few methods ( like UseOracle ) to Microsoft.EntityFrameworkCore namespace which is allready imported

File: generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs (hunk 4 above)
Comment (danielpetisme) [reply generators/server/templates/dotnetcore/src/Project/Infrastructure/DatabaseStartup.cs.ejs]: Ok don't touch it

File: generators/server/templates/dotnetcore/src/Project/Project.csproj.ejs (hunk 6 above)
Comment (danielpetisme) [reply generators/server/templates/dotnetcore/src/Project/Project.csproj.ejs]: 👍 for another PR for mongo

**Comment by danielpetisme** [comment]
Looks good to me, other stuff to add? otherwise I can merge it

**Comment by sergrdz** [comment]
It would nice to include the files for mssql, mysql and postgresql database services for docker-compose in this request

**Comment by nicolas63** [comment]
Can you wait @danielpetisme ? the inheritance of the startup is not a good practice i have to look better before merge. Yes it could be Nice if you add docker images. Nice work !

**Comment by ItsInsajd** [comment]
@nicolas63 Do you see any better way to resolve it? Rewritting the whole startup for tests would mean copy pasting most of the code, which is also far from ideal

**Comment by nicolas63** [comment]
It's ok for this solution, like jhipster tests are integration tests.
If unit tests are created in next version we will find a better solution at that time.

**Comment by danielpetisme** [comment]
@nicolas63 can you have a final look please.

**Comment by nicolas63** [comment]
Why you don't use official docker image for oracle ?

**Comment by ItsInsajd** [comment]
I tried using `oracle-database-enterprise-edition` which seems to be official, but I get error when pulling the image 

> pull access denied for oracle-database-enterprise-edition, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

This image is commercial, so I guess it can't be used without paid licence. I can leave this official image if you want, but I didn't test it. 
The image we used is also not ideal, because the db server needs some time to start up, so on first run the web app will fail at connection. We had to wait for db to start (there are logs in console stating when it's ready) and then restart the image with web app again. 

To be honest with you, I feel like oracle is used mostly in enterprise environment and I'm not sure if we can set it up to work perfectly out of the box

**Comment by ruddell** [comment]
For reference, in the main `generator-jhipster` project, we do not supply a docker-compose config for Oracle.

https://github.com/jhipster/generator-jhipster/blob/10fa9ba215ede70096d880f804b1e4a0f1fd438b/generators/server/files.js#L65-L69

**Comment by nicolas63** [comment]
Good remarks @ruddell. @ItsInsajd can you remove image for oracle and add this link in the documentation ? https://github.com/oracle/docker-images/blob/master/OracleDatabase/SingleInstance/README.md

**Comment by danielpetisme** [comment]
@nicolas63 you now have the merge permission. 🎉 
I propose to merge this PR whenever you think it's ready

**Comment by nicolas63** [comment]
yes =). I just wait oracle image supression and i merge it

**Comment by ItsInsajd** [comment]
One question, you want me to remove the whole db service from compose in case of oracle, or just the image part?

**Comment by nicolas63** [comment]
You can remove the db service in case of oracle if the user want test with oracle he can use this script (https://github.com/oracle/docker-images/tree/master/OracleDatabase/SingleInstance)

**Comment by ItsInsajd** [comment]
Done. Btw, dockerfile won't work properly for now, because of dotnetcore sdk mismatch. We didn't commit this change, because #76 fixes that

**Comment by nicolas63** [comment]
Nice work ! Thank you very much to you !

**Pull request merged**